codemp/qcommon/files.cpp

```diff
@@ -1006,15 +1063,73 @@ void FS_FCloseFile( fileHandle_t f ) {
 		if ( fsh[f].handleFiles.unique ) {
 			unzClose( fsh[f].handleFiles.file.z );
 		}
-		Com_Memset( &fsh[f], 0, sizeof( fsh[f] ) );
+		FS_ResetFileHandleData( &fsh[f] );
 		return;
 	}
 
 	// we didn't find it as a pak, so close it as a unique file
 	if (fsh[f].handleFiles.file.o) {
-		fclose (fsh[f].handleFiles.file.o);
+		if ( fsh[f].handleAsync ) {
+			// queue the file to be closed after all pending operations are completed.
+			{
+				std::lock_guard<std::mutex> l( fsh[f].writeLock );
+				fsh[f].closed = qtrue;
+			}
+			fsh[f].cv.notify_one();
+			return;
+		} else {
+			fclose (fsh[f].handleFiles.file.o);
+		}
 	}
-	Com_Memset( &fsh[f], 0, sizeof( fsh[f] ) );
+	FS_ResetFileHandleData( &fsh[f] );
+}
+
+extern void Com_PushEvent( sysEvent_t *event );
+void FS_AsyncWriterThread( fileHandle_t h ) {
+	fileHandleData_t *f = &fsh[h];
+	if ( !FS_CreatePath( f->ospath ) ) {
```

Comment:
Ugh... No threads + exception is tricky... I think if you use std::async the exception is passed along when you .get from the std::future or some such? I'm not sure...